src/main/kotlin/com/exactpro/th2/rptdataprovider/handlers/messages/MessageExtractor.kt

```diff
@@ -138,139 +99,180 @@ class MessageExtractor<B, G, RM, PM>(
 
+                for (batch: StoredGroupedMessageBatch in cradleMessageIterable) {
+                    if (externalScope.isActive && lastNotFound) {
                         val timeStart = System.currentTimeMillis()
 
-                        pipelineStatus.fetchedStart(streamName.toString())
-                        logger.trace { "batch ${batch.id.sequence} of stream $streamName with ${batch.messageCount} messages (${batch.batchSize} bytes) has been extracted" }
+                        pipelineStatus.fetchedStart(commonStreamName.toString())
+                        LOGGER.trace { "batch ${batch.firstTimestamp} of group $sessionGroup for stream $commonStreamName with ${batch.messageCount} messages (${batch.batchSize} bytes) has been extracted" }
+
+                        // TODO: this is not optimal solution (BEGIN)
+                        val isBatchOrdered = batch.messages.asSequence()
+                            .filter { it.sessionAlias == commonStreamName.name }
+                            .map { it.timestamp }
+                            .runningReduce { previous, current -> if (current < previous) null else current }
+                            .all { it != null }
+
+                        var streamSequence = if (isBatchOrdered) {
+                            batch.orderedMessages.asSequence().filter { it.sessionAlias == commonStreamName.name }
+                        } else {
+                            val orderedMessages = batch.messages
+                                .filter { it.sessionAlias == commonStreamName.name }
+                                .sortedBy { it.timestamp }
+
+                            when (request.searchDirection) {
+                                AFTER -> orderedMessages
+                                BEFORE -> orderedMessages.asReversed()
+                            }.asSequence()
+                        }
+                        // TODO: this is not optimal solution (END)
 
-                        val trimmedMessages = trimMessagesListHead(getMessagesFromBatch(batch), resumeFromId)
-                            .dropLastWhile { message ->
-                                trimMessagesListTail(message).also {
-                                    isLastMessageTrimmed = isLastMessageTrimmed || it
+                        if (firstNotFound) {
+                            streamSequence = streamSequence.dropWhile(
+                                when {
+                                    resumeFromId != null -> {
+                                        { msg -> (msg.direction != resumeFromId.stream.direction || sequenceComparator(msg.sequence, resumeFromId.sequence)).also { firstNotFound = it } }
+                                    }
+                                    else /* start != null */ -> {
+                                        { msg -> timestampComparator(msg.timestamp, start).also { firstNotFound = it } }
+                                    }

```

Comment:
issue: this code doesn't consider case when resumeFromId and start aren't null, but rpt-dp has test for this case